File: Editor/src/main/java/com/rspsi/studio/theme/StudioDrawColors.java
package com.rspsi.studio.theme;

/**
 * Dear ImGui's {@code ImDrawList} primitives (addRectFilled, addRect,
 * addText, addTriangleFilled, ...) pack color ints as {@code 0xAABBGGRR},
 * not the {@code 0xAARRGGBB} every other convention in this codebase uses -
 * a raw hex color handed to them directly comes out with red and blue
 * swapped. {@code ImGui.pushStyleColor}/{@code textColored} and the rest of
 * the style-color API do not need this: their binding already normalizes to
 * the natural {@code 0xAARRGGBB} order. Route every literal color through
 * here before it reaches an {@code ImDrawList} call.
 */
public final class StudioDrawColors {
    private StudioDrawColors() {
    }

    public static int abgr(int argb) {
        int a = (argb >>> 24) & 0xFF;
        int r = (argb >>> 16) & 0xFF;
        int g = (argb >>> 8) & 0xFF;
        int b = argb & 0xFF;
        return (a << 24) | (b << 16) | (g << 8) | r;
    }
}

File: Editor/src/main/kotlin/com/rspsi/studio/theme/StudioDrawColors.kt
package com.rspsi.studio.theme

/**
 * Converts normal ARGB colors to Dear ImGui ImDrawList's packed ABGR representation.
 *
 * ImGui style-color APIs already normalize ARGB and must not use this conversion.
 */
object StudioDrawColors {
    @JvmStatic
    fun abgr(argb: Int): Int {
        val a = (argb ushr 24) and 0xFF
        val r = (argb ushr 16) and 0xFF
        val g = (argb ushr 8) and 0xFF
        val b = argb and 0xFF
        return (a shl 24) or (b shl 16) or (g shl 8) or r
    }
}
